exchanges/kline/kline.go

```diff
@@ -0,0 +1,128 @@
+func CreateKline(trades []order.TradeHistory, interval time.Duration, p currency.Pair, a asset.Item, exchange string) (Item, error) {
+	if interval < time.Minute {
+		return Item{}, fmt.Errorf("invalid time interval: [%s]", interval)
+	}
+
+	err := validateData(trades)
+	if err != nil {
+		return Item{}, err
+	}
+
+	timeIntervalStart := trades[0].Timestamp.Truncate(interval)
+	timeIntervalEnd := trades[len(trades)-1].Timestamp
+
+	// Adds time interval buffer zones
+	var timeIntervalCache [][]order.TradeHistory
+	var OpenClose []HeartBeat
+
+	for t := timeIntervalStart; t.Before(timeIntervalEnd); t = t.Add(interval) {
+		timeBufferEnd := t.Add(interval)
+		insertionCount := 0
+
+		var zonedTradeHistory []order.TradeHistory
+		for i := 0; i < len(trades); i++ {
+			if (trades[i].Timestamp.After(t) ||
+				trades[i].Timestamp.Equal(t)) &&
+				(trades[i].Timestamp.Before(timeBufferEnd) ||
+					trades[i].Timestamp.Equal(timeBufferEnd)) {
+				zonedTradeHistory = append(zonedTradeHistory, trades[i])
+				insertionCount++
+				continue
+			}
+			trades = trades[i:]
+			break
+		}
+
+		// Insert dummy in time period when there is no price action
+		if insertionCount == 0 {
+			OpenClose = append(OpenClose, HeartBeat{Open: t, Close: timeBufferEnd})
+			timeIntervalCache = append(timeIntervalCache, []order.TradeHistory{})
+			continue
+		}
+		OpenClose = append(OpenClose, HeartBeat{Open: t, Close: timeBufferEnd})
+		timeIntervalCache = append(timeIntervalCache, zonedTradeHistory)
+	}
+
+	var candles = Item{
+		Exchange: exchange,
+		Pair:     p,
+		Asset:    a,
+		Interval: interval,
+	}
+
+	var closePriceOfLast float64
+	for x := range timeIntervalCache {
+		if len(timeIntervalCache[x]) == 0 {
+			candles.Candles = append(candles.Candles, Candle{
+				Time:  OpenClose[x].Open,
+				High:  closePriceOfLast,
+				Low:   closePriceOfLast,
+				Close: closePriceOfLast,
+				Open:  closePriceOfLast})
+			continue
+		}
+
+		var newCandle Candle
+		for y := range timeIntervalCache[x] {
+			if y == 0 {
+				newCandle.Open = timeIntervalCache[x][y].Price
+				newCandle.Time = OpenClose[x].Open
+			}
+			if y == len(timeIntervalCache[x])-1 {
+				newCandle.Close = timeIntervalCache[x][y].Price
+				closePriceOfLast = timeIntervalCache[x][y].Price
+			}
+			if newCandle.High < timeIntervalCache[x][y].Price {
+				newCandle.High = timeIntervalCache[x][y].Price
+			}
+			if newCandle.Low > timeIntervalCache[x][y].Price || newCandle.Low == 0 {
+				newCandle.Low = timeIntervalCache[x][y].Price
+			}
+			newCandle.Volume += timeIntervalCache[x][y].Amount
+		}
+		candles.Candles = append(candles.Candles, newCandle)
+	}
+	return candles, nil
+}
+
+// validatData checks for zero values on data and sorts before turning
+// converting into OHLC
+func validateData(trades []order.TradeHistory) error {
+	if len(trades) < 2 {
+		return errors.New("insufficient data")
+	}
+
+	for i := range trades {
+		if trades[i].Timestamp.IsZero() ||
+			trades[i].Timestamp.Unix() == 0 {
```

Comment:
Is there a difference between these two?

Reply:
```
func TestSomething(t *testing.T) {
	wow1 := time.Unix(0, 0)
	wow2 := time.Time{}
	fmt.Printf("Time.Unix() Is Zero %v, Unix Val %v\n", wow1.IsZero(), wow1.Unix())
	fmt.Printf("Time.Time{} Is Zero %v, Unix Val %v\n", wow2.IsZero(), wow2.Unix())
}
```
Output:
Time.Unix() Is Zero false, Unix Val 0
Time.Time{} Is Zero true, Unix Val -62135596800

I think it was a weird edge case.